### PR TITLE
Implement QM31 libfuncs

### LINKED ISSUE #1425: [2.12.0] Implement QM31 related Libfuncs
Implement the following libfuncs related to QM31:
- [QM31Concrete::BinaryOperation](https://github.com/starkware-libs/cairo/blob/96625b57abee8aca55bdeb3ecf29f82e8cea77c3/crates/cairo-lang-sierra-to-casm/src/invocations/qm31.rs#L33)
- [QM31Concrete::IsZero](https://github.com/starkware-libs/cairo/blob/96625b57abee8aca55bdeb3ecf29f82e8cea77c3/crates/cairo-lang-sierra-to-casm/src/invocations/qm31.rs#L24) 
- [QM31Concrete::Const](https://github.com/starkware-libs/cairo/blob/96625b57abee8aca55bdeb3ecf29f82e8cea77c3/crates/cairo-lang-sierra-to-casm/src/invocations/qm31.rs#L65)
- [QM31Concrete::Pack](https://github.com/starkware-libs/cairo/blob/96625b57abee8aca55bdeb3ecf29f82e8cea77c3/crates/cairo-lang-sierra-to-casm/src/invocations/qm31.rs#L81)
- [QM31Concrete::Unpack](https://github.com/starkware-libs/cairo/blob/96625b57abee8aca55bdeb3ecf29f82e8cea77c3/crates/cairo-lang-sierra-to-casm/src/invocations/qm31.rs#L110)
- [QM31Concrete::FromM31](https://github.com/starkware-libs/cairo/blob/96625b57abee8aca55bdeb3ecf29f82e8cea77c3/crates/cairo-lang-sierra-to-casm/src/invocations/qm31.rs#L28)

Consider implementing them on rust runtime with the QM31 api from types-rs ([here](https://github.com/starknet-io/types-rs/blob/main/crates/starknet-types-core/src/qm31/mod.rs))

### PR DESCRIPTION
**Pull request by DiegoCivi**
# Implement QM31 libfuncs

Implements all the QM31 related libfuncs:
- [x] [QM31Concrete::BinaryOperation](https://github.com/starkware-libs/cairo/blob/96625b57abee8aca55bdeb3ecf29f82e8cea77c3/crates/cairo-lang-sierra-to-casm/src/invocations/qm31.rs#L33)
- [x] [QM31Concrete::IsZero](https://github.com/starkware-libs/cairo/blob/96625b57abee8aca55bdeb3ecf29f82e8cea77c3/crates/cairo-lang-sierra-to-casm/src/invocations/qm31.rs#L24)
- [x] [QM31Concrete::Const](https://github.com/starkware-libs/cairo/blob/96625b57abee8aca55bdeb3ecf29f82e8cea77c3/crates/cairo-lang-sierra-to-casm/src/invocations/qm31.rs#L65)
- [x] [QM31Concrete::Pack](https://github.com/starkware-libs/cairo/blob/96625b57abee8aca55bdeb3ecf29f82e8cea77c3/crates/cairo-lang-sierra-to-casm/src/invocations/qm31.rs#L81)
- [x] [QM31Concrete::Unpack](https://github.com/starkware-libs/cairo/blob/96625b57abee8aca55bdeb3ecf29f82e8cea77c3/crates/cairo-lang-sierra-to-casm/src/invocations/qm31.rs#L110)
- [x] [QM31Concrete::FromM31](https://github.com/starkware-libs/cairo/blob/96625b57abee8aca55bdeb3ecf29f82e8cea77c3/crates/cairo-lang-sierra-to-casm/src/invocations/qm31.rs#L28)

Also added some tests and I was able to run the QM31 tests from the corelib:
```shell
% cargo run -p cairo-native-test corelib/src/test/qm31_test.cairo -s
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.27s
     Running `target/debug/cairo-native-test corelib/src/test/qm31_test.cairo -s`
running 6 tests
test qm31_test::qm31_test::test_qm31_add_and_sub ... ok (gas usage est.: 2100)
test qm31_test::qm31_test::test_qm31_mul_and_div ... ok (gas usage est.: 2500)
test qm31_test::qm31_test::test_qm31_inverse ... ok (gas usage est.: 2600)
test qm31_test::qm31_test::test_pack ... ok (gas usage est.: 5700)
test qm31_test::qm31_test::test_unpack ... ok (gas usage est.: 14800)
test qm31_test::qm31_test::test_m31_into_qm31 ... ok (gas usage est.: 13600)

test result: ok. 6 passed; 0 failed; 0 ignored; 0 filtered out;
```

Closes #1425

## Introduces Breaking Changes?

Yes/No.

- [ ] Created PR in [sequencer](https://github.com/lambdaclass/sequencer)
- [ ] Created PR in [starknet-replay](https://github.com/lambdaclass/starknet-replay)   
- [ ] Updated the `starknet-blocks.yml` workflow to use these PRs.

These PRs should be merged after this one right away, in that order.

## Checklist

- [ ] Linked to Github Issue.
- [ ] Unit tests added.
- [ ] Integration tests added.
- [ ] This change requires new documentation.
  - [ ] Documentation has been added/updated.

### CODE REVIEW

File: src/libfuncs/qm31.rs [hunk 1]
```diff
@@ -0,0 +1,712 @@
...
+    let qm31 = entry.insert_value(context, location, qm31, m31_0, 0)?;
+    let qm31 = entry.insert_value(context, location, qm31, m31_1, 1)?;
+    let qm31 = entry.insert_value(context, location, qm31, m31_2, 2)?;
+    let qm31 = entry.insert_value(context, location, qm31, m31_3, 3)?;
+
+    helper.br(entry, 0, &[qm31], location)
+}
+
+/// Generate MLIR operations for the `qm31_is_zero` libfunc.
+pub fn build_is_zero<'ctx, 'this>(
```
Comment (FrancoGiachetta) [inline src/libfuncs/qm31.rs]: I think this should be done in runtime. However, I also think there are some things that could be simplified here:
```rust
let qm31 = entry.arg(0)?;
    let m31_ty = IntegerType::new(context, 31).into();
    let qm31_ty = llvm::r#type::r#struct(context, &[m31_ty, m31_ty, m31_ty, m31_ty], false);
    let qm31_ptr =
        helper
            .init_block
            .alloca1(context, location, qm31_ty, get_integer_layout(31).align())?;
    let cond_ptr = helper.init_block.alloca1(
        context,
        location,
        IntegerType::new(context, 1).into(),
        get_integer_layout(1).align(),
    )?;
```
1. Both `cond_ptr` and `qm31_ptr` should not necessary:
 * `cond_ptr`: it is just a bit so it is ffi safe, meaning it can be passed to the function as it is.
 * `qm31_ptr`: Is a struct which is in essence a pointer. And since its fields are all `u31`s  probably is can be passed as a `&[u32; 4]` to rust.
 
2. The alloca is being performed in the init_block, which shouldn't. It can be done with:
```rust
entry.alloca1(...)?;
```

File: src/libfuncs/qm31.rs [hunk 2]
```diff
@@ -0,0 +1,712 @@
...
+            )?;
+        }
+        cairo_lang_sierra::extensions::qm31::QM31BinaryOperator::Div => {
+            let op = entry.const_int_from_type(context, location, 3, op_ty)?;
+            entry.store(context, location, op_ptr, op)?;
+            runtime_bindings_meta.libfunc_qm31_bin_op(
+                context, helper, entry, lhs_ptr, rhs_ptr, op_ptr, res_ptr, location,
+            )?;
+        }
+    };
```
Comment (FrancoGiachetta) [inline src/libfuncs/qm31.rs]: I think this match is not necessary, you could cast the enum as an `u8` (bare in mind you check it respects the ordering you are using now).

File: src/types/qm31.rs [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+///
+/// Check out [the module](self) for more info.
+pub fn build<'ctx>(
+    context: &'ctx Context,
+    _module: &Module<'ctx>,
+    _registry: &ProgramRegistry<CoreType, CoreLibfunc>,
+    _metadata: &mut MetadataStorage,
+    _info: &InfoOnlyConcreteType,
+) -> Result<Type<'ctx>> {
+    let m31 = IntegerType::new(context, 31).into();
```
Comment (JulianGCalderon) [inline src/types/qm31.rs]: Should we use 32 bits instead? It is true that we need 31 bits to store an M31, but any operation on them (like adding to m31) would require us to extend to 32 bits anyway, so perhaps its better to use 32 directly?

File: src/runtime.rs [hunk 4]
```diff
@@ -583,6 +583,101 @@ pub unsafe extern "C" fn cairo_native__libfunc__ec__ec_state_try_finalize_nz(
...
 }
 
+/// Compute `qm31_is_zero(qm31)` and store the result.
+///
+/// # Safety
+///
+/// This function is intended to be called from MLIR, deals with pointers, and is therefore
+/// definitely unsafe to use manually.
+pub unsafe extern "C" fn cairo_native__libfunc__qm31__qm31_is_zero(
+    qm31: &[[u8; 4]; 4],
```
Comment (JulianGCalderon) [inline src/runtime.rs]: Can we define a `QM31Abi` type, like `Felt252Abi`?

https://github.com/lambdaclass/cairo_native/blob/35361e287744ce95abf4528da5e82c24fc8d261b/src/starknet.rs?plain=1#L38

File: src/libfuncs/qm31.rs [hunk 5]
```diff
@@ -0,0 +1,712 @@
...
+            build_unpack(context, registry, entry, location, helper, metadata, info)
+        }
+        QM31Concrete::FromM31(info) => {
+            build_from_m31(context, registry, entry, location, helper, metadata, info)
+        }
+    }
+}
+
+/// Generate MLIR operations for the `qm31_const` libfunc.
+pub fn build_const<'ctx, 'this>(
```
[the remark this answers is not in the document]
Comment (JulianGCalderon) [reply src/libfuncs/qm31.rs]: Also, delve into the preconditions of each libfunc. Does it ensure that the `M31` values are within range, or should we take it into account?

File: src/types/qm31.rs [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+    Ok(llvm::r#type::r#struct(
+        context,
+        &[m31, m31, m31, m31],
+        false,
+    ))
```
Comment (FrancoGiachetta) [inline src/types/qm31.rs]: I would use an `llvm::r#type::array` instead. QM31 is a represents 4 `u32`s. While an struct allows it fields to be from different types, an array implies they are all from the same one. So I think the proper, and safest, type would be an array.

File: src/executor.rs [hunk 7]
```diff
@@ -433,6 +433,12 @@ fn parse_result(
             registry,
             true,
         )?),
+        CoreTypeConcrete::QM31(_) => Ok(Value::from_ptr(
+            return_ptr.to_native_assert_error("return pointer should be valid")?,
+            type_id,
+            registry,
+            true,
+        )?),
```
Comment (JulianGCalderon) [inline src/executor.rs]: I'm not sure this is the right implementation. Is this required on this PR? or can be left for a future PR?

File: src/libfuncs/qm31.rs [hunk 8]
```diff
@@ -0,0 +1,712 @@
...
+
+        let result = run_program(&program, "run_test_large_coefficients", &[]).return_value;
+        assert_eq!(
+            result,
+            Value::QM31(0x544b2fba, 0x673cff77, 0x60713d44, 0x499602d2)
+        );
+    }
+
+    #[test]
+    fn run_is_zero() {
```
Comment (JulianGCalderon) [inline src/libfuncs/qm31.rs]: The is_zero has some edge cases when an `m31` value is exactly `M31_PRIME`. We should add a test for that edge case.

File: src/runtime.rs [hunk 9]
```diff
@@ -583,6 +583,101 @@ pub unsafe extern "C" fn cairo_native__libfunc__ec__ec_state_try_finalize_nz(
...
+    cond_ptr: &mut [u8; 1],
+) {
+    let qm31 = *qm31;
+    let qm31_0 = u32::from_le_bytes(qm31[0]);
+    let qm31_1 = u32::from_le_bytes(qm31[1]);
+    let qm31_2 = u32::from_le_bytes(qm31[2]);
+    let qm31_3 = u32::from_le_bytes(qm31[3]);
+
+    let qm31 = starknet_types_core::qm31::QM31::from_coefficients(qm31_0, qm31_1, qm31_2, qm31_3);
```
[the remark this answers is not in the document]
Comment (JulianGCalderon) [reply src/runtime.rs]: I looked into this a bit. `from_coefficients` calls `weak_reduce` on each input. Is there a way to remove that operation?

File: src/libfuncs/qm31.rs (hunk 8 above)
Comment (DiegoCivi) [reply src/libfuncs/qm31.rs]: `M31_PRIME` is equal to **2147483647** while the maximum value for a cairo m31 is **2147483646** since they are of type `BoundedInt<0, 2147483646>`. So this edge case should not happen thanks to cairo.
```rust
/// The field for the Mersenne prime with `n` 31.
#[feature("bounded-int-utils")]
pub type m31 =
    core::internal::bounded_int::BoundedInt<0, 0x7ffffffe>; // 0x7ffffffe = 2147483646
```

File: src/runtime.rs [hunk 10]
```diff
@@ -583,6 +583,101 @@ pub unsafe extern "C" fn cairo_native__libfunc__ec__ec_state_try_finalize_nz(
...
+) {
+    let m31 = u32::from_le_bytes(*m31);
+    let qm31 = starknet_types_core::qm31::QM31::from_coefficients(m31, 0, 0, 0);
+    let coefficients = qm31.to_coefficients();
+
+    qm31_res[0] = coefficients.0.to_le_bytes();
+    qm31_res[1] = coefficients.1.to_le_bytes();
+    qm31_res[2] = coefficients.2.to_le_bytes();
+    qm31_res[3] = coefficients.3.to_le_bytes();
+}
```
[the remark this answers is not in the document]
Comment (DiegoCivi) [reply src/runtime.rs]: Right, done in [`798033b`](https://github.com/lambdaclass/cairo_native/pull/1429/commits/798033b6be421f4ecc96ee51a7ed936f0dd98dfe)

File: src/runtime.rs [hunk 11]
```diff
@@ -583,6 +583,101 @@ pub unsafe extern "C" fn cairo_native__libfunc__ec__ec_state_try_finalize_nz(
...
 
+/// Compute `qm31_is_zero(qm31)` and store the result.
+///
+/// # Safety
+///
+/// This function is intended to be called from MLIR, deals with pointers, and is therefore
+/// definitely unsafe to use manually.
+pub unsafe extern "C" fn cairo_native__libfunc__qm31__qm31_is_zero(
+    qm31: &[[u8; 4]; 4],
+    cond_ptr: &mut [u8; 1],
```
[the remark this answers is not in the document]
Comment (DiegoCivi) [reply src/runtime.rs]: This libfunc was moved to mlir

File: src/runtime.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (DiegoCivi) [reply src/runtime.rs]: Moved that libfunc to mlir, however I used this in the binary operations libfuncs

File: src/types.rs [hunk 12]
```diff
@@ -843,7 +845,7 @@ impl TypeBuilder for CoreTypeConcrete {
                 .is_memory_allocated(registry)?,
             CoreTypeConcrete::Coupon(_) => false,
             CoreTypeConcrete::Circuit(_) => false,
-            CoreTypeConcrete::QM31(_) => native_panic!("Implement is_memory_allocated for QM31"),
+            CoreTypeConcrete::QM31(_) => true,
```
Comment (FrancoGiachetta) [inline src/types.rs]: I think this is false. A struct in essence is a static array.

File: src/executor.rs (hunk 7 above)
Comment (JulianGCalderon) [reply src/executor.rs]: I compiled the following example

```c
// file: a.c
#include <stdint.h>
typedef struct T { int v[4]; } T;
T foo() { T var = {1,2,3,4}; return var; }
```

for both aarch64 and x86_64, and found that the value is returned through registers.

```
$ clang -target aarch64-linux-gnu -O2 a.c -S -o -
...
        mov     x0, #1                          // =0x1
        mov     x1, #3                          // =0x3
        movk    x0, #2, lsl #32
        movk    x1, #4, lsl #32
        ret
...

$ clang -target x86_64-linux-gnu -O2 a.c -S -o -
...
        movabsq $8589934593, %rax               # imm = 0x200000001
        movabsq $17179869187, %rdx              # imm = 0x400000003
        retq
...
```

File: src/libfuncs/qm31.rs [hunk 13]
```diff
@@ -0,0 +1,785 @@
...
+    let m31_0 = entry.const_int_from_type(context, location, info.w0, m31_ty)?;
+    let m31_1 = entry.const_int_from_type(context, location, info.w1, m31_ty)?;
+    let m31_2 = entry.const_int_from_type(context, location, info.w2, m31_ty)?;
+    let m31_3 = entry.const_int_from_type(context, location, info.w3, m31_ty)?;
+
+    let qm31 = entry.append_op_result(llvm::undef(qm31_ty, location))?;
+    let qm31 = entry.insert_value(context, location, qm31, m31_0, 0)?;
+    let qm31 = entry.insert_value(context, location, qm31, m31_1, 1)?;
+    let qm31 = entry.insert_value(context, location, qm31, m31_2, 2)?;
+    let qm31 = entry.insert_value(context, location, qm31, m31_3, 3)?;
```
Comment (JulianGCalderon) [inline src/libfuncs/qm31.rs]: We can use `insert_values` (plural) here

File: src/libfuncs/qm31.rs [hunk 14]
```diff
@@ -0,0 +1,785 @@
...
+    let m31_2 = entry.extract_value(context, location, qm31, m31_ty, 2)?;
+    let m31_3 = entry.extract_value(context, location, qm31, m31_ty, 3)?;
+
+    let cond = entry.append_op_result(arith::ori(m31_0, m31_1, location))?;
+    let cond = entry.append_op_result(arith::ori(cond, m31_2, location))?;
+    let cond = entry.append_op_result(arith::ori(cond, m31_3, location))?;
+
+    let k0 = entry.const_int_from_type(context, location, 0, m31_ty)?;
+    let cond =
+        entry.append_op_result(arith::cmpi(context, CmpiPredicate::Eq, k0, cond, location))?;
```
Comment (JulianGCalderon) [inline src/libfuncs/qm31.rs]: This is correct, however a bit confusing. Can you add a comment to clarify that we are actually doing: `(m31_0 | m31_1 | m31_2 | m31_3) == 0`, which is the same as checking each limb for zero.

File: src/libfuncs/qm31.rs [hunk 15]
```diff
@@ -0,0 +1,785 @@
...
+    let m31_3 = entry.arg(3)?;
+
+    let m31_ty = IntegerType::new(context, 31).into();
+    let qm31_ty = llvm::r#type::array(m31_ty, 4);
+
+    let qm31 = entry.append_op_result(llvm::undef(qm31_ty, location))?;
+    let qm31 = entry.insert_value(context, location, qm31, m31_0, 0)?;
+    let qm31 = entry.insert_value(context, location, qm31, m31_1, 1)?;
+    let qm31 = entry.insert_value(context, location, qm31, m31_2, 2)?;
+    let qm31 = entry.insert_value(context, location, qm31, m31_3, 3)?;
```
Comment (JulianGCalderon) [inline src/libfuncs/qm31.rs]: We can use `insert_values` (plural) here

File: src/libfuncs/qm31.rs [hunk 16]
```diff
@@ -0,0 +1,785 @@
...
+    helper.br(entry, 0, &[result], location)
+}
+
+/// Generate MLIR operations for the `qm31_pack` libfunc.
+///
+/// Receives four m31 and packs them into a qm31.
+///
+/// # Constraints
+///
+/// m31 are always between 0 and 2**31 - 2
```
Comment (JulianGCalderon) [inline src/libfuncs/qm31.rs]:
```suggestion
/// m31 are always between 0 and 2**31 - 2 (inclusive)
```

File: src/libfuncs/qm31.rs [hunk 17]
```diff
@@ -0,0 +1,785 @@
...
+                let b = QM31Trait::new(0x499602d2, 0x544b2fba, 0x673cff77, 0x60713d44);
+                let c = QM31Trait::new(0x1de1328d, 0x3b882f32, 0x47ae3cbc, 0x2a074017);
+                b + c
+            }
+
+            fn run_a_plus_c() -> qm31 {
+                let a = QM31Trait::new(0x544b2fba, 0x673cff77, 0x60713d44, 0x499602d2);
+                let c = QM31Trait::new(0x1de1328d, 0x3b882f32, 0x47ae3cbc, 0x2a074017);
+                a + c
+            }
```
Comment (JulianGCalderon) [inline src/libfuncs/qm31.rs]: Could you add a test case for `(PRIME - 1) + 1`, and assert that the return value is `0`?

**Review by JulianGCalderon — CHANGES_REQUESTED** [review]

File: src/libfuncs/qm31.rs [hunk 18]
```diff
@@ -0,0 +1,785 @@
...
+
+    let m31_ty = IntegerType::new(context, 31).into();
+    let qm31_ty = llvm::r#type::array(m31_ty, 4);
+    let k0 = entry.const_int_from_type(context, location, 0, m31_ty)?;
+
+    let qm31 = entry.append_op_result(llvm::undef(qm31_ty, location))?;
+    let qm31 = entry.insert_value(context, location, qm31, m31, 0)?;
+    let qm31 = entry.insert_value(context, location, qm31, k0, 1)?;
+    let qm31 = entry.insert_value(context, location, qm31, k0, 2)?;
+    let qm31 = entry.insert_value(context, location, qm31, k0, 3)?;
```
[the remark this answers is not in the document]
Comment (DiegoCivi) [reply src/libfuncs/qm31.rs]: Done in [`2e551c4`](https://github.com/lambdaclass/cairo_native/pull/1429/commits/2e551c47c45353fa30a2075d1287b85ecb686f5b)

File: src/metadata/runtime_bindings.rs [hunk 19]
```diff
@@ -547,6 +568,44 @@ impl RuntimeBindingsMeta {
         ))
     }
 
+    #[allow(clippy::too_many_arguments)]
+    pub fn libfunc_qm31_bin_op<'c, 'a>(
```
[the remark this answers is not in the document]
Comment (DiegoCivi) [reply src/metadata/runtime_bindings.rs]: Done in [`daccb07`](https://github.com/lambdaclass/cairo_native/pull/1429/commits/daccb07fb601d4778c7c2fbd38e16928a77124ef) and in [`c41ad6f`](https://github.com/lambdaclass/cairo_native/pull/1429/commits/c41ad6f245156472dbf92745ff6d83eff2f0a3f6)

File: src/executor.rs [hunk 20]
```diff
@@ -433,6 +433,21 @@ fn parse_result(
             registry,
             true,
         )?),
+        CoreTypeConcrete::QM31(_) => match return_ptr {
+            Some(ptr) => Ok(Value::from_ptr(ptr, type_id, registry, true)?),
+            None => {
+                #[cfg(target_arch = "x86_64")]
+                return Err(Error::ParseAttributeError);
```
Comment (JulianGCalderon) [inline src/executor.rs]: I would add a comment with something like: "on x86_64, a QM31 value is always returned by memory, therefore making this branch is unreachable on that architecture".

**Review by JulianGCalderon — COMMENTED** [review]
We should reenable the corelib QM31 tests. See https://github.com/lambdaclass/cairo_native/blob/35361e287744ce95abf4528da5e82c24fc8d261b/corelib.patch

File: src/metadata/runtime_bindings.rs [hunk 21]
```diff
@@ -547,6 +569,59 @@ impl RuntimeBindingsMeta {
...
 
+    /// Register binary operation function if necessary and invoke it.
+    /// The operation depends on the `op` argument which could indicate:
+    /// - Add operation
+    /// - Sub operation
+    /// - Mul operation
+    /// - Div operation
+    ///
+    /// Operation is done between `lhs_ptr` and `rhs_ptr` while the result is stored
+    /// in `res_ptr`.
```
[the remark this answers is not in the document]
Comment (DiegoCivi) [reply src/metadata/runtime_bindings.rs]: Mmm I wouldn't say outdated, the operation is still using those pointers and the result is being stored in the res_ptr which is now created in the function and is not a parameter. I added the mention about the pointers here [`b7b0bd0`](https://github.com/lambdaclass/cairo_native/pull/1429/commits/b7b0bd07ba346fc3c6d151b9dc3b719d2b778d89)

File: src/executor.rs (hunk 20 above)
Comment (DiegoCivi) [reply src/executor.rs]: Done in [`b7b0bd0`](https://github.com/lambdaclass/cairo_native/pull/1429/commits/b7b0bd07ba346fc3c6d151b9dc3b719d2b778d89)

**Review by gabrielbosio — CHANGES_REQUESTED** [review]

File: src/runtime.rs [hunk 22]
```diff
@@ -583,6 +585,100 @@ pub unsafe extern "C" fn cairo_native__libfunc__ec__ec_state_try_finalize_nz(
...
+    // We can use this way of creating the QM31 since we already know from cairo that the
+    // coefficients will never be more than 31 bits wide
+    let lhs = QM31(Degree4ExtensionField::const_from_coefficients(
+        lhs[0], lhs[1], lhs[2], lhs[3],
+    ));
+    let rhs = QM31(Degree4ExtensionField::const_from_coefficients(
+        rhs[0], rhs[1], rhs[2], rhs[3],
+    ));
+
+    // SAFETY: The only possible error is if rhs is zero. However, in the QM31 division libfunc, the divisor
```
Comment (gabrielbosio) [inline src/runtime.rs]:
```suggestion
    // SAFETY: An error would be triggered here only if rhs is zero. However, in the QM31 division libfunc, the divisor
```

File: src/libfuncs/qm31.rs [hunk 23]
```diff
@@ -0,0 +1,1116 @@
...
+/// it will manage (QM31 or M31). It either receives 2 qm31 or 2 m31 (which are represented
+/// as bounded ints)
+///
+/// # Cairo Signature
+/// ```cairo
+/// // qm31
+/// fn qm31_add(a: qm31, b: qm31) -> qm31 nopanic;
+/// fn qm31_sub(a: qm31, b: qm31) -> qm31 nopanic;
+/// fn qm31_mul(a: qm31, b: qm31) -> qm31 nopanic;
+/// fn qm31_mul(a: qm31, b: qm31) -> qm31 nopanic;
```
Comment (gabrielbosio) [inline src/libfuncs/qm31.rs]:
```suggestion
/// fn qm31_div(a: qm31, b: NonZero<qm31>) -> qm31 nopanic;
```

File: src/libfuncs/qm31.rs
Comment (gabrielbosio) [inline src/libfuncs/qm31.rs]: How about using `ProgramRegistry::build_type` like in [felt252.rs](https://github.com/lambdaclass/cairo_native/blob/18529434ac21d6cbb93058e8928011a1716fae99/src/libfuncs/felt252.rs#L69)?

Comment (DiegoCivi) [reply src/libfuncs/qm31.rs]: I like that, done in [`cf81a58`](https://github.com/lambdaclass/cairo_native/pull/1429/commits/cf81a58819ca5b37c532376642482055e69e1b8c)

File: src/libfuncs/qm31.rs [hunk 24]
```diff
@@ -0,0 +1,1116 @@
...
+    context: &'ctx Context,
+    entry: &'this Block<'ctx>,
+    location: Location<'ctx>,
+    helper: &LibfuncHelper<'ctx, 'this>,
+) -> Result<()> {
+    let lhs_value = entry.arg(0)?;
+    let rhs_value = entry.arg(1)?;
+
+    let res = entry.append_op_result(arith::subi(lhs_value, rhs_value, location))?;
+    let prime = entry.const_int(context, location, M31_PRIME, 31)?;
```
[the remark this answers is not in the document]
Comment (DiegoCivi) [reply src/libfuncs/qm31.rs]: Since the add can have an overflow we need that extra bit, however in the substraction it is not necessary. We can have it for consistency but we would need to extend bothe the lhs and the rhs and then truncate the result to 31 bits again.

File: src/metadata/runtime_bindings.rs [hunk 25]
```diff
@@ -547,6 +569,59 @@ impl RuntimeBindingsMeta {
...
 
+    /// Register QM31 binary operation function if necessary and invoke it.
+    /// The operation depends on the `op` argument which could indicate:
+    /// - Add operation
+    /// - Sub operation
+    /// - Mul operation
+    /// - Div operation
+    ///
+    /// Executes the operation on the `QM31` values referenced by `lhs_ptr` and `rhs_ptr`,
+    /// and stores the resulting `QM31` in `res_ptr`.
```
Comment (JulianGCalderon) [inline src/metadata/runtime_bindings.rs]: I wouldn't mention `res_ptr`, its an implementation detail. The user should only care that this function takes two pointers to an `QM31`, and returns  a `QM31` value.

File: src/metadata/runtime_bindings.rs [hunk 26]
```diff
@@ -547,6 +569,59 @@ impl RuntimeBindingsMeta {
...
+    ///
+    /// Returns a opaque pointer as the result.
```
Comment (JulianGCalderon) [inline src/metadata/runtime_bindings.rs]: It doesn't return an opaque pointer, but a QM31 value, because of the `load` at line 622

File: src/executor.rs (hunk 20 above)
Comment (JulianGCalderon) [reply src/executor.rs]: A QM31 is not bigger than an u128, but its returned as a pointer regardless. I would remove that part of the comment as its probably more complex than that.

File: src/metadata/runtime_bindings.rs (hunk 25 above)
Comment (DiegoCivi) [reply src/metadata/runtime_bindings.rs]: Done in [`6b5e32e`](https://github.com/lambdaclass/cairo_native/pull/1429/commits/6b5e32ef31c9768037c5d0cc0fadf26a5f9fcec1)

File: src/runtime.rs [hunk 27]
```diff
@@ -617,6 +713,25 @@ thread_local! {
     };
 }
 
+// TODO: This is already implemented on types-rs but there is no release
+// that contains it. It should be deleted when bumping to a new version
+// and use the .to_coefficients() method from QM31 instead.
+pub fn to_representative_coefficients(qm31: QM31) -> [u32; 4] {
```
Comment (FrancoGiachetta) [inline src/runtime.rs]: I would `qm31_to_representative_coefficients` just to be sure this function is associated to that type.

**Review by FrancoGiachetta — DISMISSED** [review]

**Review by JulianGCalderon — DISMISSED** [review]
LGTM!

**Review by FrancoGiachetta — APPROVED** [review]
